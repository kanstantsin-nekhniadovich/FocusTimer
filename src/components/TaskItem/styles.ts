import { StyleSheet } from 'react-native';
import { Colors, Typography } from '@styles';

export const styles = StyleSheet.create({
  item: {
    width: '100%',
    flexDirection: 'row',
    justifyContent: 'space-between',
    paddingVertical: 4,
    paddingHorizontal: 6,
    alignItems: 'center',
    height: 65,
    borderBottomWidth: StyleSheet.hairlineWidth,
    borderColor: Colors.darkBlue,
    backgroundColor: Colors.alabaster,
  },
  titleHolder: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  title: {
    ...Typography.subtitle,
    fontSize: 16,
    color: Colors.prussianBlue,
    textTransform: 'capitalize',
    marginLeft: 15,
    maxWidth: 180,
  },
  firstItem: {
    borderTopWidth: StyleSheet.hairlineWidth,
    borderColor: Colors.darkBlue,
  },
  newTaskLabel: {
    ...Typography.text,
    color: Colors.prussianBlue,
  },
  timeHolder: {
    width: 65,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  remainingTime: {
    ...Typography.primary,
    color: Colors.silverChaliceLight,
    textTransform: 'lowercase',
  },
  confirmMessage: {
    ...Typography.message,
    lineHeight: 20,
    color: Colors.white,
  },
  swipeAction: {
    flex: 1,
    alignItems: 'center',
    flexDirection: 'row',
    paddingHorizontal: 10,
    backgroundColor: Colors.darkBlue,
  },
  swipeRightAction: {
    justifyContent: 'flex-end',
  },
});
