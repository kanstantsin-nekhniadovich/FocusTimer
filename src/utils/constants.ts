export const TOKEN = 'token';
export const FIREBASE_TOKEN_KEY = 'firebaseToken';
export const USER_SKIPPED_LOGIN_FLOW_KEY = 'isUserSkippedLoginFlow';

export const DEFAULT_WORK_TIME_IN_MINUTES = 25;
export const DEFAULT_BREAK_TIME_IN_MINUTES = 5;

export enum Status {
  TODO = 'TODO',
  INPROGRESS = 'INPROGRESS',
  COMPLETED = 'COMPLETED',
}
