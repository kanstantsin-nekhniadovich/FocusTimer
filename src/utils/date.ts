const MILLISECONDS_IN_SECOND = 1000;
const MILLISECONDS_IN_MINUTE = 60 * MILLISECONDS_IN_SECOND;
const MILLISECONDS_IN_HOUR = 60 * MILLISECONDS_IN_MINUTE;

const paddingStart = (value: number) => value.toString().padStart(2, '0');

const convertMillisecondsInHours = (milliseconds: number) => Math.floor(milliseconds / MILLISECONDS_IN_HOUR);

const convertMillisecondsInMinutes = (milliseconds: number, rounded: boolean = false) =>
  rounded
    ? Math.round(milliseconds / MILLISECONDS_IN_MINUTE)
    : Math.floor(milliseconds / MILLISECONDS_IN_MINUTE);

export const formatMillisecondsToMinutes = (milliseconds: number) => `${convertMillisecondsInMinutes(milliseconds)} min`;

export const convertMillisecondsToTimeFormat = (milliseconds: number) => {
  const hours = convertMillisecondsInHours(milliseconds);
  const restForMinutes = milliseconds - hours * MILLISECONDS_IN_HOUR;

  const minutes = convertMillisecondsInMinutes(restForMinutes);
  const restForSeconds = restForMinutes - minutes * MILLISECONDS_IN_MINUTE;

  const seconds = restForSeconds / MILLISECONDS_IN_SECOND;

  return hours
    ? `${paddingStart(hours)}:${paddingStart(minutes)}:${paddingStart(seconds)}`
    : `${paddingStart(minutes)}:${paddingStart(seconds)}`;
};

export const convertMinutesInMilliseconds = (minutes: number) => minutes * MILLISECONDS_IN_MINUTE;
