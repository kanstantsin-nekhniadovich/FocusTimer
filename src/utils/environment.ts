import Constants from 'expo-constants';

import { isDefined } from './isDefined';

enum Environments {
  DEV = 'development',
  PROD = 'production',
}

const getEnvironment = (): Environments => {
  const releaseChannel = Constants.manifest.releaseChannel;

  if (isDefined(releaseChannel) && releaseChannel === Environments.PROD) {
    return Environments.PROD;
  }

  return Environments.DEV;
};


export const isProdEnv = (): boolean => getEnvironment() === Environments.PROD;
export const isDevEnv = (): boolean => getEnvironment() === Environments.DEV;
