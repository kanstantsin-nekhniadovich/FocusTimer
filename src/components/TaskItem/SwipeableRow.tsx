import React from 'react';
import { Swipeable, RectButton } from 'react-native-gesture-handler';
import { Colors } from '@styles';

import { isDefined } from '../../utils/isDefined';
import { Delete } from '../icons';

import { styles } from './styles';

interface Props {
  // onSwipeableOpen: () => void;
  onActionClick: () => void;
  onSwipeableLeftWillOpen: () => void;
  onSwipeableRightWillOpen: () => void;
  onSwipeableWillClose: () => void;
}

export const SwipeableRow: React.FC<Props> =
  React.memo(({ onActionClick, onSwipeableLeftWillOpen, onSwipeableRightWillOpen, onSwipeableWillClose, children }) => {
    const ref = React.useRef<Nullable<Swipeable>>(null);

    const close = () => {
      if (!isDefined(ref.current)) {
        return;
      }

      ref.current.close();
    };

    const handleActionClick = React.useCallback(() => {
      onActionClick();
      close();
    }, [onActionClick]);

    const handleSwipeableOpen = React.useCallback(() => {
      // onSwipeableOpen();
      onActionClick();
    }, [onActionClick]);

    const renderRightActions = () => {
      return (
        <RectButton style={{ ...styles.swipeAction, ...styles.swipeRightAction }} onPress={handleActionClick}>
          <Delete color={Colors.white} width={24} />
        </RectButton>
      );
    };

    const renderLeftActions = () => {
      return (
        <RectButton style={styles.swipeAction} onPress={handleActionClick}>
          <Delete color={Colors.white} width={24} />
        </RectButton>
      );
    };

    return (
      <Swipeable
        ref={ref}
        friction={2}
        leftThreshold={50}
        rightThreshold={50}
        overshootFriction={8}
        onSwipeableOpen={handleSwipeableOpen}
        renderRightActions={renderRightActions}
        renderLeftActions={renderLeftActions}
        onSwipeableLeftWillOpen={onSwipeableLeftWillOpen}
        onSwipeableRightWillOpen={onSwipeableRightWillOpen}
        onSwipeableWillClose={onSwipeableWillClose}
        onSwipeableWillOpen={() => { console.log('asdfsdfsdfsd'); }}
      >
        {children}
      </Swipeable>
    );
  });
