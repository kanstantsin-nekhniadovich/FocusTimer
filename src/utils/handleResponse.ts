import { ResponseStatus } from '../utils/constants';

export const handleResponse = <T>(response: NormalizedResponse<T>) =>
  <S, F>(successHandler: (res: SuccessResponse<T>) => S, failureHandler: (res: FailureResponse) => F) => {
    if (response.status === ResponseStatus.SUCCESS) {
      return successHandler(response);
    }

    return failureHandler(response);
  };
