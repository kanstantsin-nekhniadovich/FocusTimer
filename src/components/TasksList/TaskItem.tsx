import { Task } from '@typings';
import React from 'react';
import { Text, View } from 'react-native';
import { Colors } from '@styles';

import { formatMillisecondsToMinutes } from '../../utils/date';
import { Status } from '../../utils/constants';
import { CircleProgressBar } from '../common';
import { Time, Tick } from '../icons';

import { itemStyles as styles } from './styles';

interface Props {
  task: Task;
}

export const TaskItem: React.FC<Props> = ({ task }) => {
  const isNewTask = task.status === Status.TODO && task.workTime === task.remainingTime;
  const isCompletedTask = task.status === Status.COMPLETED && task.remainingTime === 0;

  const progress = isNewTask ? 0 : (task.remainingTime / task.workTime) * 100;

  return (
    <View style={styles.item}>
      <View style={styles.titleHolder}>
        {isCompletedTask
          ? <Tick />
          : <CircleProgressBar size={36} strokeWidth={3} progress={progress} progressColor={Colors.darkBlue} />
        }
        <Text style={styles.title} numberOfLines={1}>{task.title}</Text>
      </View>
      {isNewTask
        ? <Text style={styles.newTaskLabel}>New Task</Text>
        : (!isCompletedTask &&
          <View style={styles.timeHolder}>
            <Time />
            <Text style={styles.remainingTime}>{formatMillisecondsToMinutes(task.remainingTime)}</Text>
          </View>)
      }
    </View>
  );
};
