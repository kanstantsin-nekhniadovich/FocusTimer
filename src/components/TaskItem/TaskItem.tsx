import { Task } from '@typings';
import React from 'react';
import { useDispatch } from 'react-redux';
import { Text, View, Animated } from 'react-native';
import { Colors, Typography } from '@styles';

import { deleteTaskRequest } from '../../ducks';
import { formatMillisecondsToMinutes } from '../../utils/date';
import { Status } from '../../utils/constants';
import { CircleProgressBar, ConfirmModal } from '../common';
import { Time, Tick } from '../icons';
import { SwipeableRow } from './SwipeableRow';

import { styles } from './styles';

interface Props {
  task: Task;
}

export const TaskItem: React.FC<Props> = ({ task }) => {
  const dispatch = useDispatch();
  const [isRemoveTaskModalOpened, setIsRemoveTaskModalOpened] = React.useState(false);

  const closeModal = (close: () => void) => () => {
    setIsRemoveTaskModalOpened(false);
    close();
  };

  const showModal = React.useCallback(() => {
    setIsRemoveTaskModalOpened(true);
  }, []);

  const removeTask = (close: () => void) => () => {
    dispatch(deleteTaskRequest(task.id));
    close();
  };

  const isNewTask = task.status === Status.TODO && task.workTime === task.remainingTime;
  const isCompletedTask = task.status === Status.COMPLETED && task.remainingTime === 0;

  const progress = isNewTask ? 0 : (task.remainingTime / task.workTime) * 100;

  return (
    <SwipeableRow onSwipeableOpenAction={showModal}>
      {({ close }) => (
        <Animated.View style={styles.item}>
          <View style={styles.titleHolder}>
            {isCompletedTask
              ? <Tick />
              : <CircleProgressBar size={36} strokeWidth={3} progress={progress} progressColor={Colors.darkBlue} />
            }
            <Text style={styles.title} numberOfLines={1}>{task.title}</Text>
          </View>
          {isNewTask
            ? <Text style={styles.newTaskLabel}>New Task</Text>
            : (!isCompletedTask &&
          <View style={styles.timeHolder}>
            <Time />
            <Text style={styles.remainingTime}>{formatMillisecondsToMinutes(task.remainingTime)}</Text>
          </View>)
          }
          <ConfirmModal
            title="Remove task"
            isVisible={isRemoveTaskModalOpened}
            onCancel={closeModal(close)}
            onOk={removeTask(close)}
            okText="Remove"
            cancelText="Close"
            closeOnClickOutside={false}
            messageContent={<Text style={styles.confirmMessage}>
          Are you sure that you want to remove <Text style={Typography.subtitleLarge}>{task.title}</Text> task?
            </Text>}
          />
        </Animated.View>
      )}
    </SwipeableRow>
  );
};
