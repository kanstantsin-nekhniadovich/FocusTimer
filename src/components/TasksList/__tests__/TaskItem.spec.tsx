import React from 'react';
import { render } from '@testing-library/react-native';

import { TaskItem } from '../TaskItem';
import { Status } from '../../../utils/constants';

jest.mock('expo-constants', () => ({
  manifest: {
    extra: {
      firebaseApiKey: ''
    },
  }
}));

describe('<TaskItem />', () => {
  it('should render TaskItem component with task in TODO', () => {
    const task = {
      id: '123',
      title: 'Test task',
      cyclesCount: 2,
      workTime: 1500000,
      breakTime: 20000,
      status: Status.TODO,
      remainingTime: 1500000,
      currentCycle: 1,
      projectId: '1',
    };

    const { toJSON } = render(<TaskItem task={task}/>);

    expect(toJSON()).toMatchSnapshot();
  });

  it('should render TaskItem component with task in INPROGRESS', () => {
    const task = {
      id: '123',
      title: 'Test task',
      cyclesCount: 2,
      workTime: 1500000,
      breakTime: 20000,
      status: Status.INPROGRESS,
      remainingTime: 500000,
      currentCycle: 1,
      projectId: '2',
    };

    const { toJSON } = render(<TaskItem task={task}/>);

    expect(toJSON()).toMatchSnapshot();
  });

  it('should render TaskItem component with COMPLETED task', () => {
    const task = {
      id: '123',
      title: 'Test task',
      cyclesCount: 2,
      workTime: 1500000,
      breakTime: 20000,
      status: Status.COMPLETED,
      remainingTime: 0,
      currentCycle: 1,
      projectId: '2',
    };

    const { toJSON } = render(<TaskItem task={task}/>);

    expect(toJSON()).toMatchSnapshot();
  });
});
