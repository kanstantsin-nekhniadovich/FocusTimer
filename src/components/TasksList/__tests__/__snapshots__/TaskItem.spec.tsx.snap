// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TaskItem /> should render TaskItem component with COMPLETED task 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "height": 36,
      "justifyContent": "space-between",
      "marginBottom": 20,
      "paddingVertical": 4,
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <RNSVGSvgView
      align="xMidYMid"
      bbHeight={36}
      bbWidth={36}
      fill="none"
      focusable={false}
      height={36}
      meetOrSlice={0}
      minX={0}
      minY={0}
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          Object {
            "flex": 0,
            "height": 36,
            "width": 36,
          },
        ]
      }
      vbHeight={36}
      vbWidth={36}
      width={36}
    >
      <RNSVGGroup
        fill={null}
        propList={
          Array [
            "fill",
          ]
        }
      >
        <RNSVGPath
          d="M18 0C8.064 0 0 8.064 0 18C0 27.936 8.064 36 18 36C27.936 36 36 27.936 36 18C36 8.064 27.936 0 18 0ZM14.4 27L5.4 18L7.938 15.462L14.4 21.906L28.062 8.244L30.6 10.8L14.4 27Z"
          fill={
            Array [
              1,
              "paint0_linear",
            ]
          }
          propList={
            Array [
              "fill",
            ]
          }
        />
        <RNSVGDefs>
          <RNSVGLinearGradient
            gradient={
              Array [
                0,
                -16274536,
                1,
                -14820979,
              ]
            }
            gradientTransform={null}
            gradientUnits={1}
            name="paint0_linear"
            x1="36"
            x2="0"
            y1="18"
            y2="18"
          />
        </RNSVGDefs>
      </RNSVGGroup>
    </RNSVGSvgView>
    <Text
      numberOfLines={1}
      style={
        Object {
          "color": "#001f5a",
          "fontFamily": "DMSans_500Medium",
          "fontSize": 16,
          "letterSpacing": 0,
          "lineHeight": 20,
          "marginLeft": 15,
          "maxWidth": 180,
          "textTransform": "capitalize",
        }
      }
    >
      Test task
    </Text>
  </View>
</View>
`;

exports[`<TaskItem /> should render TaskItem component with task in INPROGRESS 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "height": 36,
      "justifyContent": "space-between",
      "marginBottom": 20,
      "paddingVertical": 4,
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <RNSVGSvgView
      bbHeight={36}
      bbWidth={36}
      focusable={false}
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          Object {
            "height": 36,
            "transform": Array [
              Object {
                "rotate": "-90deg",
              },
            ],
            "width": 36,
          },
          Object {
            "flex": 0,
            "height": 36,
            "width": 36,
          },
        ]
      }
    >
      <RNSVGGroup
        matrix={
          Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ]
        }
      >
        <RNSVGCircle
          cx={18}
          cy={18}
          propList={
            Array [
              "stroke",
              "strokeWidth",
            ]
          }
          r={16.5}
          stroke={4292335847}
          strokeWidth={3}
        />
        <RNSVGCircle
          cx={18}
          cy={18}
          propList={
            Array [
              "stroke",
              "strokeWidth",
              "strokeDasharray",
              "strokeDashoffset",
            ]
          }
          r={16.5}
          stroke={4280297654}
          strokeDasharray={
            Array [
              103.67255756846318,
              103.67255756846318,
            ]
          }
          strokeDashoffset={69.11503837897546}
          strokeWidth={3}
        />
      </RNSVGGroup>
    </RNSVGSvgView>
    <Text
      numberOfLines={1}
      style={
        Object {
          "color": "#001f5a",
          "fontFamily": "DMSans_500Medium",
          "fontSize": 16,
          "letterSpacing": 0,
          "lineHeight": 20,
          "marginLeft": 15,
          "maxWidth": 180,
          "textTransform": "capitalize",
        }
      }
    >
      Test task
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "width": 65,
      }
    }
  >
    <RNSVGSvgView
      align="xMidYMid"
      bbHeight={16}
      bbWidth={16}
      fill="none"
      focusable={false}
      height={16}
      meetOrSlice={0}
      minX={0}
      minY={0}
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          Object {
            "flex": 0,
            "height": 16,
            "width": 16,
          },
        ]
      }
      vbHeight={16}
      vbWidth={16}
      width={16}
    >
      <RNSVGGroup
        fill={null}
        propList={
          Array [
            "fill",
          ]
        }
      >
        <RNSVGPath
          d="M7.9935 1.33325C4.3135 1.33325 1.3335 4.31992 1.3335 7.99992C1.3335 11.6799 4.3135 14.6666 7.9935 14.6666C11.6802 14.6666 14.6668 11.6799 14.6668 7.99992C14.6668 4.31992 11.6802 1.33325 7.9935 1.33325ZM8.00016 13.3333C5.0535 13.3333 2.66683 10.9466 2.66683 7.99992C2.66683 5.05325 5.0535 2.66659 8.00016 2.66659C10.9468 2.66659 13.3335 5.05325 13.3335 7.99992C13.3335 10.9466 10.9468 13.3333 8.00016 13.3333Z"
          fill={4288980132}
          propList={
            Array [
              "fill",
            ]
          }
        />
        <RNSVGPath
          d="M8.3335 4.66675H7.3335V8.66675L10.8335 10.7667L11.3335 9.94675L8.3335 8.16675V4.66675Z"
          fill={4288980132}
          propList={
            Array [
              "fill",
            ]
          }
        />
      </RNSVGGroup>
    </RNSVGSvgView>
    <Text
      style={
        Object {
          "color": "#a4a4a4",
          "fontFamily": "Roboto_500Medium",
          "fontSize": 14,
          "letterSpacing": 0.14,
          "lineHeight": 24,
          "textTransform": "lowercase",
        }
      }
    >
      8 min
    </Text>
  </View>
</View>
`;

exports[`<TaskItem /> should render TaskItem component with task in TODO 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "height": 36,
      "justifyContent": "space-between",
      "marginBottom": 20,
      "paddingVertical": 4,
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <RNSVGSvgView
      bbHeight={36}
      bbWidth={36}
      focusable={false}
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          Object {
            "height": 36,
            "transform": Array [
              Object {
                "rotate": "-90deg",
              },
            ],
            "width": 36,
          },
          Object {
            "flex": 0,
            "height": 36,
            "width": 36,
          },
        ]
      }
    >
      <RNSVGGroup
        matrix={
          Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ]
        }
      >
        <RNSVGCircle
          cx={18}
          cy={18}
          propList={
            Array [
              "stroke",
              "strokeWidth",
            ]
          }
          r={16.5}
          stroke={4292335847}
          strokeWidth={3}
        />
        <RNSVGCircle
          cx={18}
          cy={18}
          propList={
            Array [
              "stroke",
              "strokeWidth",
              "strokeDasharray",
              "strokeDashoffset",
            ]
          }
          r={16.5}
          stroke={4280297654}
          strokeDasharray={
            Array [
              103.67255756846318,
              103.67255756846318,
            ]
          }
          strokeDashoffset={103.67255756846318}
          strokeWidth={3}
        />
      </RNSVGGroup>
    </RNSVGSvgView>
    <Text
      numberOfLines={1}
      style={
        Object {
          "color": "#001f5a",
          "fontFamily": "DMSans_500Medium",
          "fontSize": 16,
          "letterSpacing": 0,
          "lineHeight": 20,
          "marginLeft": 15,
          "maxWidth": 180,
          "textTransform": "capitalize",
        }
      }
    >
      Test task
    </Text>
  </View>
  <Text
    style={
      Object {
        "color": "#001f5a",
        "fontFamily": "Roboto_400Regular",
        "fontSize": 14,
        "letterSpacing": -0.3,
        "lineHeight": 18,
      }
    }
  >
    New Task
  </Text>
</View>
`;
